perun-core/src/main/java/cz/metacentrum/perun/core/blImpl/GroupsManagerBlImpl.java

```diff
@@ -508,21 +508,44 @@ public void moveGroup(PerunSession sess, Group destinationGroup, Group movingGro
+			//check if there is union between destination group and moving group
+			if(groupsManagerImpl.isOneWayRelationBetweenGroups(destinationGroup, movingGroup)) {
+				throw new GroupMoveNotAllowedException("There is already group union between moving group: " + movingGroup + " and destination group: " + destinationGroup + ".", movingGroup, destinationGroup);
+			}
 
-			try {
-				//Removing indirect members from old position just if moving group is not top level group
-				if (movingGroup.getParentGroupId() != null) {
-					removeRelationMembers(sess, groupsManagerImpl.getGroupById(sess,movingGroup.getParentGroupId()), members, movingGroup.getId());
+			//We have to remove old group relation, if moving group is not top level group
+			if(movingGroup.getParentGroupId() != null){
+				try {
+					removeGroupUnion(sess, getParentGroup(sess, movingGroup), movingGroup,true );
+				} catch (GroupRelationDoesNotExist e) {
+					//that should never happened
+					throw new InternalErrorException("Group relation does not exists between group " + movingGroup + "and its parent group.");
+				} catch (GroupRelationCannotBeRemoved e) {
+					//that should never happened
+					throw new InternalErrorException("Group relation cannot be removed between group " + movingGroup + "and its parent group.");
+				} catch (ParentGroupNotExistsException e) {
+					//That should never happened
+					throw new InternalErrorException("Parent group does not exists for group " + movingGroup);
+				} catch (GroupNotExistsException e) {
+					throw new InternalErrorException(e);
 				}
-				//Adding indirect members to new position
-				addRelationMembers(sess, groupsManagerImpl.getGroupById(sess, destinationGroup.getId()), members, movingGroup.getId());
+			}
 
-			} catch (WrongReferenceAttributeValueException | NotGroupMemberException | AlreadyMemberException | WrongAttributeValueException e ) {
-				throw new InternalErrorException("Some operation was not allowed during processing members relations", e);
+			//we have to create new group relation
+			try {
+				createGroupUnion(sess, destinationGroup, movingGroup, true);
+			} catch (GroupRelationAlreadyExists e) {
+				//that should noever happened
+				throw new InternalErrorException("Group relation already exists between destination group "  + destinationGroup + " and moving group " + movingGroup + ".");
+			} catch (GroupRelationNotAllowed e) {
+				//that should never happened
+				throw new InternalErrorException("Group relation cannot be created between destination group "  + destinationGroup + " and moving group " + movingGroup + ".");
 			} catch (GroupNotExistsException e) {
-				throw new InternalErrorException("Some group does not exists...", e);
+				throw new InternalErrorException(e);
+			} catch (WrongAttributeValueException e) {
+				throw new InternalErrorException(e);
+			} catch (WrongReferenceAttributeValueException e) {
+				throw new InternalErrorException(e);
```

Comment:
If you have the same throw from more than one exception, you should merge them together (if there is no difference).